File: libraries/php/tests/ClientTest.php
<?php

declare(strict_types=1);

namespace Voidhash\Tests;

use PHPUnit\Framework\TestCase;
use Voidhash\Exception\ApiException;
use Voidhash\VoidhashClient;
use Voidhash\Webhooks;
use Voidhash\Generated\Core\Model\CreateSecretKeyBodyJsonEncoding;
use Voidhash\Generated\Core\Model\SetPersonAttributesBodyJsonEncoding;

final class ClientTest extends TestCase
{
    private StubHttpClient $http;
    private VoidhashClient $client;

    protected function setUp(): void
    {
        $this->http = new StubHttpClient();
        $this->client = VoidhashClient::create('vh_sk_test', ['httpClient' => $this->http]);
    }

    public function testGetByDistinctIdSendsAuthHeaderAndDecodes(): void
    {
        $this->http->queueJson(200, ['personId' => 'per_1', 'distinctId' => 'user-123', 'email' => null, 'name' => null]);

        $person = $this->client->persons->getByDistinctId('user-123');

        self::assertSame('per_1', $person?->getPersonId());
        $request = $this->http->requests[0];
        self::assertSame('/api/v1/persons/by-distinct-id/user-123', $request->getUri()->getPath());
        self::assertSame('vh_sk_test', $request->getHeaderLine('x-secret-key'));
    }

    public function testCaptureSendsThePublishableKeyAsTheBodyTokenWhenConfigured(): void
    {
        $client = VoidhashClient::create('vh_sk_test', [
            'httpClient' => $this->http,
            'publishableKey' => 'vh_pk_test',
        ]);
        $this->http->queueJson(202, ['accepted' => 1, 'rejected' => 0]);

        $result = $client->eventCapture->capture([
            'event' => 'note_created',
            'distinctId' => 'user-123',
        ]);

        self::assertSame(['accepted' => 1, 'rejected' => 0], $result);

        $request = $this->http->requests[0];
        self::assertSame('/i/v1/capture', $request->getUri()->getPath());
        // The secret key is the credential; the body token only mirrors what a
        // browser SDK would send.
        self::assertSame('vh_sk_test', $request->getHeaderLine('x-secret-key'));

        $body = json_decode((string) $request->getBody(), true);
        self::assertSame('note_created', $body['event']);
        self::assertSame('user-123', $body['distinct_id']);
        self::assertSame('vh_pk_test', $body['token']);
        self::assertNotEmpty($body['uuid']);
        self::assertNotEmpty($body['sent_at']);
        // Both must be JSON objects; `[]` is rejected with a 400.
        self::assertStringContainsString('"context":{}', (string) $request->getBody());
        self::assertStringContainsString('"properties":{}', (string) $request->getBody());
    }

    public function testBatchSendsThePublishableKeyAsTheBodyTokenWhenConfigured(): void
    {
        $client = VoidhashClient::create('vh_sk_test', [
            'httpClient' => $this->http,
            'publishableKey' => 'vh_pk_test',
        ]);
        $this->http->queueJson(202, ['accepted' => 1, 'rejected' => 0]);

        $client->eventCapture->batch([['event' => 'note_created', 'distinctId' => 'user-123']]);

        $body = json_decode((string) $this->http->requests[0]->getBody(), true);
        self::assertSame('vh_pk_test', $body['token']);
    }

    public function testSetAttributesPostsTraitsForTheNamedPerson(): void
    {
        $this->http->queueJson(200, ['personId' => 'per_1', 'distinctId' => 'user-123', 'email' => null, 'name' => null]);

        $person = $this->client->persons->setAttributes(
            (new SetPersonAttributesBodyJsonEncoding())
                ->setDistinctId('user-123')
                ->setTraits(['plan' => 'pro', 'notes_created' => 3]),
        );

        self::assertSame('per_1', $person?->getPersonId());
        $request = $this->http->requests[0];
        self::assertSame('/api/v1/persons/attributes', $request->getUri()->getPath());
        self::assertSame('vh_sk_test', $request->getHeaderLine('x-secret-key'));

        $body = json_decode((string) $request->getBody(), true);
        self::assertSame('user-123', $body['distinctId']);
        self::assertSame(['plan' => 'pro', 'notes_created' => 3], $body['traits']);
    }

    public function testErrorMappingCarriesStatusAndTag(): void
    {
        $this->http->queueJson(404, ['_tag' => 'Api/PersonNotFoundError', 'id' => 'per_missing']);

        try {
            $this->client->persons->get('per_missing');
            self::fail('expected ApiException');
        } catch (ApiException $exception) {
            self::assertSame(404, $exception->getStatus());
            self::assertSame('Api/PersonNotFoundError', $exception->getTag());
        }
    }

    public function testHasActivePerkBySlugAndById(): void
    {
        $perks = fn (): array => [
            ['id' => 'perk_free', 'name' => 'Free', 'projectId' => 'prj_1', 'slug' => 'free'],
            ['id' => 'perk_pro', 'name' => 'Pro', 'projectId' => 'prj_1', 'slug' => 'pro'],
        ];
        // perks list (once per hasActivePerk call with a slug selector)
        $this->http->queueJson(200, $perks());
        // person lookup
        $this->http->queueJson(200, ['personId' => 'per_1', 'distinctId' => 'user-1']);
        // entitlements
        $this->http->queueJson(200, [
            'grants' => [[
                'perkId' => 'perk_pro',
                'status' => 'active',
                'expiresAt' => null,
                'source' => 'subscription',
                'sourceId' => null,
                'sourcePersonId' => 'per_1',
            ]],
        ]);
        // second round for the free-perk lookup
        $this->http->queueJson(200, $perks());
        $this->http->queueJson(200, ['personId' => 'per_1', 'distinctId' => 'user-1']);
        $this->http->queueJson(200, ['grants' => []]);

        $active = $this->client->persons->entitlements->hasActivePerk([
            'distinctId' => 'user-1',
            'perkSlug' => 'pro',
        ]);
        self::assertTrue($active);

        $inactive = $this->client->persons->entitlements->hasActivePerk([
            'distinctId' => 'user-1',
            'perkSlug' => 'free',
        ]);
        self::assertFalse($inactive);
    }

    public function testHasActivePerkRequiresExactlyOneSelector(): void
    {
        $this->expectException(ApiException::class);
        $this->client->persons->entitlements->hasActivePerk(['distinctId' => 'user-1']);
    }

    public function testUnknownPersonResolvesToFalseForHasActivePerk(): void
    {
        $this->http->queueJson(404, ['_tag' => 'Api/PersonNotFoundError', 'id' => 'ghost']);

        $active = $this->client->persons->entitlements->hasActivePerk([
            'distinctId' => 'ghost',
            'perkId' => 'perk_pro',
        ]);
        self::assertFalse($active);
    }

    public function testApiKeysCreateSendsBodyAndDecodesRawKey(): void
    {
        $this->http->queueJson(200, [
            'apiKey' => ['id' => 'key_1', 'name' => 'ci', 'projectId' => 'prj_1', 'maskedKey' => 'vh_sk_***'],
            'rawKey' => 'vh_sk_raw',
        ]);

        $params = (new CreateSecretKeyBodyJsonEncoding())->setName('ci')->setProjectId('prj_1');
        $created = $this->client->apiKeys->create($params);

        self::assertSame('vh_sk_raw', $created?->getRawKey());
        $request = $this->http->requests[0];
        self::assertSame('/api/v1/api-keys', $request->getUri()->getPath());
    }

    public function testCaptureSendsSecretKeyAndFullSnakeCaseBody(): void
    {
        $this->http->queueJson(202, ['accepted' => 1, 'rejected' => 0]);

        $this->client->eventCapture->capture([
            'event' => 'paywall_viewed',
            'distinctId' => 'user_123',
        ]);

        $request = $this->http->requests[0];
        self::assertSame('/i/v1/capture', $request->getUri()->getPath());
        self::assertSame('ingest.voidhash.com', $request->getUri()->getHost());
        self::assertSame('vh_sk_test', $request->getHeaderLine('x-secret-key'));

        $raw = (string) $request->getBody();
        $body = json_decode($raw, true);
        self::assertSame('paywall_viewed', $body['event']);
        self::assertSame('user_123', $body['distinct_id']);
        self::assertNotSame('', $body['uuid']);
        self::assertMatchesRegularExpression('/^[0-9a-f-]{36}$/', $body['uuid']);
        self::assertMatchesRegularExpression('/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\+00:00$/', $body['sent_at']);
        self::assertArrayNotHasKey('token', $body);
        self::assertArrayNotHasKey('session_id', $body);
        self::assertArrayNotHasKey('timestamp', $body);
        // empty maps must be JSON objects, not arrays
        self::assertStringContainsString('"properties":{}', $raw);
        self::assertStringContainsString('"context":{}', $raw);
    }

    public function testCapturePreservesCallerSuppliedFields(): void
    {
        $this->http->queueJson(202, ['accepted' => 1, 'rejected' => 0]);

        $this->client->eventCapture->capture([
            'uuid' => '018f6d2e-4c3a-7b1d-9e5f-2a8c1b0d4e6f',
            'event' => 'purchase_completed',
            'distinctId' => 'user_123',
            'properties' => ['plan' => 'pro'],
            'context' => ['app_version' => '1.2.3'],
            'sessionId' => 'sess_1',
            'timestamp' => new \DateTimeImmutable('2026-08-22T12:00:00+00:00'),
        ]);

        $body = json_decode((string) $this->http->requests[0]->getBody(), true);
        self::assertSame('018f6d2e-4c3a-7b1d-9e5f-2a8c1b0d4e6f', $body['uuid']);
        self::assertSame(['plan' => 'pro'], $body['properties']);
        self::assertSame(['app_version' => '1.2.3'], $body['context']);
        self::assertSame('sess_1', $body['session_id']);
        self::assertSame('2026-08-22T12:00:00+00:00', $body['timestamp']);
        self::assertArrayNotHasKey('token', $body);
    }

    public function testBatchPostsEnvelopeWithPerEventFields(): void
    {
        $this->http->queueJson(202, ['accepted' => 2, 'rejected' => 0]);

        $result = $this->client->eventCapture->batch([
            ['event' => 'a', 'distinctId' => 'user_1'],
            ['event' => 'b', 'distinctId' => 'user_2', 'properties' => ['k' => 'v']],
        ]);

        self::assertSame(['accepted' => 2, 'rejected' => 0], $result);

        $request = $this->http->requests[0];
        self::assertSame('/i/v1/batch', $request->getUri()->getPath());
        self::assertSame('vh_sk_test', $request->getHeaderLine('x-secret-key'));

        $raw = (string) $request->getBody();
        $body = json_decode($raw, true);
        self::assertCount(2, $body['events']);
        self::assertArrayNotHasKey('token', $body);
        self::assertArrayNotHasKey('sent_at', $body['events'][0]);
        self::assertSame('user_1', $body['events'][0]['distinct_id']);
        self::assertNotSame('', $body['events'][0]['uuid']);
        self::assertNotSame($body['events'][0]['uuid'], $body['events'][1]['uuid']);
        self::assertSame(['k' => 'v'], $body['events'][1]['properties']);
        // Nested events keep empty maps as JSON objects too; `[]` is a 400.
        self::assertStringContainsString('"context":{},"properties":{}', $raw);
        self::assertMatchesRegularExpression('/^\d{4}-\d{2}-\d{2}T/', $body['sent_at']);
    }

    public function testCaptureMapsIngestErrors(): void
    {
        $this->http->queueJson(401, ['_tag' => 'unauthorized', 'code' => 'unauthorized', 'error' => 'invalid key']);

        try {
            $this->client->eventCapture->capture(['event' => 'a', 'distinctId' => 'user_1']);
            self::fail('expected ApiException');
        } catch (ApiException $exception) {
            self::assertSame(401, $exception->getStatus());
        }
    }

    public function testSecretKeyHeaderOverrideIsRejectedCaseInsensitively(): void
    {
        $this->expectException(ApiException::class);
        VoidhashClient::create('vh_sk_test', ['headers' => ['x-Secret-Key' => 'vh_sk_other']]);
    }

    public function testEmptySecretKeyIsRejected(): void
    {
        $this->expectException(ApiException::class);
        VoidhashClient::create('  ');
    }
}

File: libraries/rust/tests/client_test.rs
//! Unit tests for the hand-written DX layer, run against a stub HTTP server
//! (wiremock) so the exact wire behavior is exercised.

use serde_json::json;
use voidhash::VoidhashClient;

async fn test_client() -> (wiremock::MockServer, VoidhashClient) {
    let server = wiremock::MockServer::start().await;
    let client = VoidhashClient::builder()
        .secret_key("vh_sk_test")
        .base_url(server.uri())
        .build()
        .expect("client builds");
    (server, client)
}

#[tokio::test]
async fn get_person_by_distinct_id_sends_auth_header_and_decodes() {
    let (server, client) = test_client().await;

    wiremock::Mock::given(wiremock::matchers::method("GET"))
        .and(wiremock::matchers::path(
            "/api/v1/persons/by-distinct-id/user-123",
        ))
        .respond_with(
            wiremock::ResponseTemplate::new(200).set_body_json(json!({
                "personId": "per_1",
                "distinctId": "user-123",
                "email": null,
                "name": null,
            })),
        )
        .mount(&server)
        .await;

    let person = client
        .persons()
        .get_by_distinct_id("user-123")
        .await
        .expect("person resolves");

    assert_eq!(person.person_id, "per_1");
    assert_eq!(person.distinct_id, "user-123");

    let requests = server.received_requests().await.expect("requests recorded");
    assert_eq!(requests.len(), 1);
    let auth_header = requests[0]
        .headers
        .iter()
        .find(|(name, _)| name.as_str() == "x-secret-key")
        .map(|(_, value)| String::from_utf8_lossy(value.as_ref()).into_owned())
        .unwrap_or_default();
    assert_eq!(auth_header, "vh_sk_test");
}

#[tokio::test]
async fn error_mapping_carries_status_and_tag() {
    let (server, client) = test_client().await;

    wiremock::Mock::given(wiremock::matchers::method("GET"))
        .and(wiremock::matchers::path("/api/v1/persons/per_missing"))
        .respond_with(
            wiremock::ResponseTemplate::new(404)
                .set_body_json(json!({"_tag": "Api/PersonNotFoundError", "id": null})),
        )
        .mount(&server)
        .await;

    let error = client
        .persons()
        .get("per_missing")
        .await
        .expect_err("expected an error");

    assert!(error.is_not_found());
    assert_eq!(
        error.to_string(),
        "voidhash api error 404: Api/PersonNotFoundError"
    );
}

#[tokio::test]
async fn has_active_perk_by_slug_and_by_id() {
    let (server, client) = test_client().await;

    // Perks list is hit once per slug lookup.
    for _ in 0..2 {
        wiremock::Mock::given(wiremock::matchers::method("GET"))
            .and(wiremock::matchers::path("/api/v1/perks"))
            .respond_with(wiremock::ResponseTemplate::new(200).set_body_json(vec![
                json!({"id": "perk_free", "name": "Free", "projectId": "prj_1", "slug": "free"}),
                json!({"id": "perk_pro", "name": "Pro", "projectId": "prj_1", "slug": "pro"}),
            ]))
            .mount(&server)
            .await;
    }
    for _ in 0..2 {
        wiremock::Mock::given(wiremock::matchers::method("GET"))
            .and(wiremock::matchers::path(
                "/api/v1/persons/by-distinct-id/user-1",
            ))
            .respond_with(
                wiremock::ResponseTemplate::new(200).set_body_json(json!({
                    "personId": "per_1", "distinctId": "user-1",
                    "email": null, "name": null,
                })),
            )
            .mount(&server)
            .await;
        wiremock::Mock::given(wiremock::matchers::method("GET"))
            .and(wiremock::matchers::path("/api/v1/persons/per_1/entitlements"))
            .respond_with(
                wiremock::ResponseTemplate::new(200).set_body_json(json!({
                    "grants": [{
                        "perkId": "perk_pro",
                        "status": "active",
                        "expiresAt": null,
                        "source": "subscription",
                        "sourceId": null,
                        "sourcePersonId": "per_1",
                    }],
                })),
            )
            .mount(&server)
            .await;
    }

    let active = client
        .persons()
        .has_active_perk("user-1", None, Some("pro"))
        .await
        .expect("slug lookup succeeds");
    assert!(active);

    let inactive = client
        .persons()
        .has_active_perk("user-1", None, Some("free"))
        .await
        .expect("slug lookup succeeds");
    assert!(!inactive);

    let by_id = client
        .persons()
        .has_active_perk("user-1", Some("perk_pro"), None)
        .await
        .expect("id lookup succeeds");
    assert!(by_id);
}

#[tokio::test]
async fn unknown_person_resolves_to_false_for_has_active_perk() {
    let (server, client) = test_client().await;

    wiremock::Mock::given(wiremock::matchers::method("GET"))
        .and(wiremock::matchers::path(
            "/api/v1/persons/by-distinct-id/ghost",
        ))
        .respond_with(
            wiremock::ResponseTemplate::new(404)
                .set_body_json(json!({"_tag": "Api/PersonNotFoundError", "id": null})),
        )
        .mount(&server)
        .await;

    let active = client
        .persons()
        .has_active_perk("ghost", Some("perk_pro"), None)
        .await
        .expect("unknown persons are not errors");
    assert!(!active);
}

#[tokio::test]
async fn has_active_perk_requires_exactly_one_selector() {
    let (_server, client) = test_client().await;

    let error = client
        .persons()
        .has_active_perk("user-1", None, None)
        .await
        .expect_err("expected a configuration error");
    assert!(matches!(error, voidhash::Error::Request(_)));

    let error = client
        .persons()
        .has_active_perk("user-1", Some("a"), Some("b"))
        .await
        .expect_err("expected a configuration error");
    assert!(matches!(error, voidhash::Error::Request(_)));
}

#[tokio::test]
async fn empty_secret_key_is_rejected() {
    let result = VoidhashClient::builder().build();
    assert!(matches!(result, Err(voidhash::Error::Request(_))));
}

#[tokio::test]
async fn capture_sends_the_ingest_contract_with_the_publishable_key() {
    let server = wiremock::MockServer::start().await;
    let client = VoidhashClient::builder()
        .secret_key("vh_sk_test")
        .publishable_key("vh_pk_test")
        .ingest_url(server.uri())
        .build()
        .expect("client builds");

    wiremock::Mock::given(wiremock::matchers::method("POST"))
        .and(wiremock::matchers::path("/i/v1/capture"))
        .respond_with(
            wiremock::ResponseTemplate::new(202)
                .set_body_json(json!({ "accepted": 1, "rejected": 0 })),
        )
        .mount(&server)
        .await;

    let result = client
        .event_capture()
        .capture(&voidhash::Event::new("user-123", "note_created"))
        .await
        .expect("capture succeeds");

    assert_eq!(result.accepted, 1);
    assert_eq!(result.rejected, 0);

    let request = &server.received_requests().await.expect("requests")[0];
    let body: serde_json::Value = serde_json::from_slice(&request.body).expect("json body");
    assert_eq!(body["event"], "note_created");
    assert_eq!(body["distinct_id"], "user-123");
    assert_eq!(body["token"], "vh_pk_test");
    assert!(body["uuid"].as_str().is_some_and(|id| !id.is_empty()));
    assert!(body["sent_at"].as_str().is_some_and(|at| !at.is_empty()));
    // Both must be JSON objects; `[]` or `null` is rejected with a 400.
    assert!(body["context"].is_object());
    assert!(body["properties"].is_object());
    // Ingest authenticates on the body token; the secret key must not leak
    // onto this origin.
    assert!(request.headers.get("x-secret-key").is_none());
}

#[tokio::test]
async fn capture_without_a_publishable_key_is_rejected() {
    let (_server, client) = test_client().await;

    let error = client
        .event_capture()
        .capture(&voidhash::Event::new("user-123", "note_created"))
        .await
        .expect_err("capture requires a publishable key");

    assert!(matches!(error, voidhash::Error::Request(_)));
}

#[tokio::test]
async fn set_attributes_posts_traits_for_the_named_person() {
    let (server, client) = test_client().await;

    wiremock::Mock::given(wiremock::matchers::method("POST"))
        .and(wiremock::matchers::path("/api/v1/persons/attributes"))
        .respond_with(
            wiremock::ResponseTemplate::new(200).set_body_json(json!({
                "personId": "per_1",
                "distinctId": "user-123",
                "email": null,
                "name": null,
            })),
        )
        .mount(&server)
        .await;

    let traits = json!({ "plan": "pro", "notes_created": 3 })
        .as_object()
        .expect("object")
        .clone();
    let person = client
        .persons()
        .set_attributes(&voidhash::PersonAttributes::new("user-123", traits))
        .await
        .expect("set_attributes succeeds");

    assert_eq!(person.person_id, "per_1");

    let request = &server.received_requests().await.expect("requests")[0];
    let body: serde_json::Value = serde_json::from_slice(&request.body).expect("json body");
    assert_eq!(body["distinctId"], "user-123");
    assert_eq!(body["traits"]["plan"], "pro");
    assert_eq!(body["traits"]["notes_created"], 3);
}
